lib/browser-specific.js

```diff
@@ -0,0 +1,247 @@
+        }
+        browserSubtests[smallestIdx].next();
+        continue;
+      }
+
+      // The iterators are all aligned at the same subtest, so score it!
+      //
+      // NOTE: There actually (rarely) exist distinct subtests with the same
+      // name in the data, usually because of unprintable characters. This can
+      // influence the result, as we will squash such subtests and may even
+      // mismatch results (i.e. if some browser has results for one
+      // duplicate-named subtest but not another).
+      //
+      // Overall the impact is minor; it at most affects a fraction of a single
+      // test, so less than 1 point of the final score per affected test.
+      denominator += 1;
+
+      let failed = [];
+      for (let i = 0; i < browserSubtests.length; i++) {
+        // TODO: Determine exact criteria for a BSF at a subtest level.
+        if (browserSubtests[i].currentValue().status == 'FAIL') {
+          failed.push(i);
+        }
+      }
+      if (failed.length == 1) {
+        counts[failed[0]] += 1;
+      }
+
+      browserSubtests.forEach(s => s.next());
+    }
+
+    for (let i = 0; i < counts.length; i++) {
+      if (counts[i] > 0) {
+        scores[i] += counts[i] / denominator;
+      }
+    }
+  }
+}
+
+// Walks a set of trees, one per browser, scoring them for browser-specific
+// failures of tests in the trees.
+function walkTrees(browserTrees, scores) {
+  // First deal with any tests that are at this level of the tree.
+  const browserTests = browserTrees.map(tree => new IteratorHelper(tree.tests));
+  while (browserTests.every(tests => tests.hasNext())) {
+    const firstTests = browserTests[0];
+
+    // If the tests are all the same object (which happens due to the caching in
+    // lib/results.js), we can just skip them; it is impossible for there to be
+    // browser-specific failures.
+    if (browserTests.every(t => t.currentValue() == firstTests.currentValue())) {
+      browserTests.forEach(t => t.next());
+      continue;
+    }
+
+    if (browserTests.every(t => t.currentKey() == firstTests.currentKey())) {
+      scoreTest(browserTests.map(t => t.currentValue()), scores);
+      browserTests.forEach(t => t.next());
+      continue;
+    }
+
+        // Our iterators are not pointing at the same test; find the earliest
+    // iterator and move it forward.
+    let smallestKey = firstTests.currentKey();
+    let smallestIdx = 0;
+    for (let i = 1; i < browserTests.length; i++) {
+      if (browserTests[i].currentKey() < smallestKey) {
+        smallestKey = browserTests[i].currentKey();
+        smallestIdx = i;
+      }
+    }
+    browserTests[smallestIdx].next();
+  }
+
+  // Now recurse into subtrees.
+  const browserSubtrees = browserTrees.map(tree => new IteratorHelper(tree.trees));
+  while (browserSubtrees.every(subtree => subtree.hasNext())) {
+    const firstTree = browserSubtrees[0];
+
+    // If the subtrees are all the same object (which happens due to the caching
```

Comment:
It seems a bit suspect that this and moving the iterator appears twice, can't that be inside the recursive call?

Reply:
I'm not sure I follow; it happens for different parts of the object. The form of the tree object is:

```
{
  trees: {
    'css': {
        trees: {
            'css-animations': { ... },
            'css-fonts': { ... },
        },
        tests: { }
    }
  },
  tests: {
    'a.html': {
      status: 'OK',
      subtests: [ ... ],
    },
    'b.html': {
      status: 'PASS',
    }
  }
}
```

So the `trees` and `tests` objects are two separate concepts at each level, and we have to do the sorted-iterator-walk and object-equality-check for each. It's not clear to me what improvement you are thinking of?

Reply:
Oh, right. That `IteratorHelper` makes the two cases look the same makes this code look more redundant than it is.